cmd/src/actions_exec.go

```diff
@@ -401,14 +401,49 @@ func prepareAction(ctx context.Context, action Action) error {
 // have been pulled from or pushed to a registry. See
 // https://windsock.io/explaining-docker-image-ids/ under "A Final Twist" for a good
 // explanation.
-func getDockerImageContentDigest(ctx context.Context, image string) (string, error) {
+func getDockerImageContentDigest(ctx context.Context, image string, logger *actionLogger) (string, error) {
 	// TODO!(sqs): is image id the right thing to use here? it is NOT the
 	// digest. but the digest is not calculated for all images (unless they are
 	// pulled/pushed from/to a registry), see
 	// https://github.com/moby/moby/issues/32016.
 	out, err := exec.CommandContext(ctx, "docker", "image", "inspect", "--format", "{{.Id}}", "--", image).CombinedOutput()
 	if err != nil {
-		return "", fmt.Errorf("error inspecting docker image (try `docker pull %q` to fix this): %s", image, bytes.TrimSpace(out))
+		// An error could mean the image just didn't exist, so we need to pull it first.
+		logger.Infof("Pulling image %q...\n", image)
+		pullCmd := exec.CommandContext(ctx, "docker", "image", "pull", image)
+		stdout, err := pullCmd.StdoutPipe()
+		if err != nil {
+			return "", fmt.Errorf("error pulling docker image %q: %s", image, err)
+		}
+		stderr, err := pullCmd.StderrPipe()
+		if err != nil {
+			return "", fmt.Errorf("error pulling docker image %q: %s", image, err)	
+		}
+		outScanner := bufio.NewScanner(stdout)
+		go func() {
+			for outScanner.Scan() {
+				logger.Infof("%s\n", outScanner.Text())
+			}
+		}()
+		errScanner := bufio.NewScanner(stderr)
+		go func() {
+			for errScanner.Scan() {
+				logger.Warnf("%s\n", errScanner.Text())
+			}
+		}()
```

Comment:
How are these goroutines terminated?

Reply:
As far as I understood it it should terminate after EOF. But not a 100% sure, the docs aren't super clear about it https://golang.org/pkg/text/scanner/#Scanner.Scan